pkg/selector/selector.go

```diff
@@ -130,77 +130,14 @@ func (itf Selector) Save() error {
-// OutputInstanceTypes() in the next major version.
-func (itf Selector) Filter(filters Filters) ([]string, error) {
-	outputFn := InstanceTypesOutputFn(outputs.SimpleInstanceTypeOutput)
-	output, _, err := itf.FilterWithOutput(filters, outputFn)
-	return output, err
-}
-
-// FilterVerbose accepts a Filters struct which is used to select the available instance types
-// matching the criteria within Filters and returns a list instanceTypeInfo
-//
-// Deprecated: This function will be replaced with GetFilteredInstanceTypes() in the next
-// major version.
-func (itf Selector) FilterVerbose(filters Filters) ([]*instancetypes.Details, error) {
-	instanceTypeInfoSlice, err := itf.rawFilter(filters)
-	if err != nil {
-		return nil, err
-	}
-	instanceTypeInfoSlice, _ = itf.truncateResults(filters.MaxResults, instanceTypeInfoSlice)
-	return instanceTypeInfoSlice, nil
-}
-
-// FilterWithOutput accepts a Filters struct which is used to select the available instance types
-// matching the criteria within Filters and returns a list of strings based on the custom outputFn
-//
-// Deprecated: This function will be replaced with GetFilteredInstanceTypes() and
-// OutputInstanceTypes() in the next major version.
-func (itf Selector) FilterWithOutput(filters Filters, outputFn InstanceTypesOutput) ([]string, int, error) {
-	instanceTypeInfoSlice, err := itf.rawFilter(filters)
-	if err != nil {
-		return nil, 0, err
-	}
-	instanceTypeInfoSlice, numOfItemsTruncated := itf.truncateResults(filters.MaxResults, instanceTypeInfoSlice)
-	output := outputFn.Output(instanceTypeInfoSlice)
-	return output, numOfItemsTruncated, nil
-}
-
-func (itf Selector) truncateResults(maxResults *int, instanceTypeInfoSlice []*instancetypes.Details) ([]*instancetypes.Details, int) {
-	if maxResults == nil {
-		return instanceTypeInfoSlice, 0
-	}
-	upperIndex := *maxResults
-	if *maxResults > len(instanceTypeInfoSlice) {
-		upperIndex = len(instanceTypeInfoSlice)
-	}
-	return instanceTypeInfoSlice[0:upperIndex], len(instanceTypeInfoSlice) - upperIndex
-}
-
-// AggregateFilterTransform takes higher level filters which are used to affect multiple raw filters in an opinionated way.
-func (itf Selector) AggregateFilterTransform(filters Filters) (Filters, error) {
-	transforms := []FiltersTransform{
-		TransformFn(itf.TransformBaseInstanceType),
-		TransformFn(itf.TransformFlexible),
-		TransformFn(itf.TransformForService),
-	}
-	var err error
-	for _, transform := range transforms {
-		filters, err = transform.Transform(filters)
-		if err != nil {
-			return filters, err
-		}
+// FilterInstanceTypes accepts a Filters struct which is used to select the available instance types
+// matching the criteria within Filters and returns the detailed specs of matching instance types
+func (itf Selector) FilterInstanceTypes(filters Filters) ([]*instancetypes.Details, error) {
+	// refresh OD and Spot pricing caches
+	if err := hydrateCaches(itf); err != nil {
```

Comment:
This shouldn't be called on every invocation of `FilterInstanceTypes`. The reason for the complex logic in main w/ cache hydration is because this process takes quite a while and is only used if pricing data is needed. The only time pricing data is needed is if the output mode is `table-wide` or a pricing related filter is being used.

Reply:
The reason I moved cache hydration to occur every time `FilterInstanceTypes` is called is so that users of instance selector as a library can have access to OD and spot pricing whenever they filter instance types. This also allows sorting by OD and spot price to occur in output formats that aren't `table-wide`.

Reply:
The problem is that this makes the lib pretty unusable because of the amount of data that needs to be fetched and stored. Most consumers of instance-selector via a lib are not using the pricing data. It needs to be smarter about when to take the hit and hydrate the pricing caches. You could change the pricing cache to only hydrate when pricing filters are passed and then when `table-wide` is used, just set the pricing filters to min:0 max: 1<<31 if not filled.

Reply:
Ah ok I understand. I see how fetching prices always might be deviating from the main use case of instance selector and will worsen the user experience. I could create logic similar to what is currently used in `main.go` to hydrate caches, which hydrates caches only when `table-wide` is used and when the pricing flag is set, and place it in `FilterInstanceTypes` instead so that library users that set the pricing flag can also get pricing information.

Reply:
I have changed `FilterInstanceTypes` to only refresh the appropriate usage class pricing cache if the pricing filters are set (same logic that was previously in `main`). I have also ensured that if the `table-wide` format is requested from the CLI, both caches will be refreshed (done in `main`).